foundry/smb3parse/util/code_edit.py

```diff
@@ -0,0 +1,95 @@
+from dataclasses import dataclass
+from abc import ABC, abstractmethod
+from typing import TypeVar, Generic
+
+from foundry.smb3parse.util.rom import Rom
+
+D = TypeVar('D')
+
+@dataclass
+class CodeEdit(ABC, Generic[D]):
+    """ Represents an area to edit to the ROM code.  
+
+    This class handles reading and writing the data of a specific code area for
+    possible code edits.  It will also check the ROM to see if the target
+    modification area has shifted due to other code modifications.
+
+    This is an abstract class.  Implementing classes will need to define the
+    read and write routines based on the type of code edit that needs to be
+    done.  The read/write routines use the Generic data type 'D' for their
+    operations.
+
+    Generic type 'D' is the abstract representation of the edit value and not
+    the literal data values (though they could be the same in some cases).  
+    See read/write for more details.
+
+    Parameters:
+        rom: Rom - the ROM to edit
+        address: int - the address of the edit
+        length: int - the length of the edit
+        prefix: int - the ROM data just before the edit.  This is used for
+            checking the validity of the target location.  This is useful for
+            when a code edit has been made that has shifted the code and
```

Comment:
Very interesting.  Another thing to consider is the header of the file.  If the change is in the last two banks of the ROM, then this would always be shifted.

Reply:
I made a note of this future improvement in the docs to look at the header and do a shift if the ROM is expanded to improve the likelihood of a successful edit on expanded ROMs, but for now, I am happy with how it is.  It will work on non-expanded ROM, and prevent an edit to expanded ROM.  It should not be able to corrupt expanded ROMs, it just doesn't work with them but also the failure is put in a warning box so I think that's fair for now.